### DISCUSSION TITLE
How to read JSON field

### DISCUSSION THREAD

**Discussion by ahmed-adly-khalil** [ClickHouse/ClickHouse#46686] [Q&A] [answered]
Hello, 
I'm having the below queries to create a queue table from NATS JetStream and move the data to an actual table via materialized view, the messages coming from NATS are JSON and everything works fine for first level attributes in the JSON object, but i can't read nested attributes

## queries 
```
CREATE TABLE payments_queue (amount UInt64, customer String) ENGINE = NATS 
    SETTINGS nats_url = '127.0.0.1:5222',
             nats_subjects = 'payments',
             nats_format = 'JSONEachRow',
             nats_token = 'token',
             date_time_input_format = 'best_effort';


 CREATE TABLE payments (amount UInt64, customer String)
    ENGINE = MergeTree() ORDER BY amount;


CREATE MATERIALIZED VIEW payments_consumer TO payments
    AS SELECT amount,  customer FROM stripe_payments_queue;


select * from payments;
```

and here is an example of the JSON coming from NATS 

```
{"amount":1500,"customer":{"id":123, .... 
```

I want to map `customer.id` to `customer` in payments table 

Thanks

**Comment by marcelbirkner** [comment, marked as answer]
Hi, 

did you have a look at https://clickhouse.com/docs/en/guides/developer/working-with-json/json-other-approaches/#using-materialized-views

Here is an example that works for me for parsing JSON using MV.

```
CREATE TABLE queue (
  key UInt64, 
  value String
)
ENGINE = MergeTree() 
ORDER BY key;

CREATE TABLE payments (
  amount UInt64, 
  customerId UInt64,
  message String
)
ENGINE = MergeTree()
ORDER BY customerId;
```

Create materialized view:

```
CREATE MATERIALIZED VIEW payments_mv TO payments AS
SELECT JSONExtractString(JSONExtractString(value, 'customer'), 'id') AS customerId,
       toUInt32(JSONExtractInt(value, 'amount')) AS amount,
       value AS message       
FROM queue;
```

Insert JSON data to `queue`:

```
INSERT INTO queue VALUES ('1', '{"amount":1500,"customer":{"id":100}}');
INSERT INTO queue VALUES ('2', '{"amount":2500,"customer":{"id":200}}');
INSERT INTO queue VALUES ('3', '{"amount":3500,"customer":{"id":300}}');
INSERT INTO queue VALUES ('4', '{"amount":4500,"customer":{"id":400}}');
```

Read data:

```
SELECT * FROM payments_mv;
```

<img width="892" alt="image" src="https://user-images.githubusercontent.com/558256/220587591-9b455fc3-6b4a-4576-a491-1706a500e2c0.png">

**Reply by ahmed-adly-khalil** [reply]
thanks so much, i'm digging into the docs to learn more 👍

**Reply by candlerb** [reply]
> did you have a look at https://clickhouse.com/docs/en/guides/developer/working-with-json/json-other-approaches/#using-materialized-views

FYI, that URL now redirects to https://clickhouse.com/docs/en/integrations/data-formats/json/overview#using-materialized-views so thank you for putting a working example in the discussion.
